src/Security/RolesDownloadStrategy.php

```diff
@@ -35,18 +36,38 @@ class RolesDownloadStrategy implements DownloadStrategyInterface
     protected $security;
 
     /**
-     * @var TranslatorInterface
+     * @var LegacyTranslatorInterface|TranslatorInterface
      */
     protected $translator;
 
     /**
      * @param AuthorizationCheckerInterface $security
      * @param string[]                      $roles
      */
-    public function __construct(TranslatorInterface $translator, $security, array $roles = [])
+    public function __construct(object $translator, $security, array $roles = [])
     {
+        if ($translator instanceof LegacyTranslatorInterface) {
+            @trigger_error(sprintf(
+                'Passing other type than "%s" as argument 1 to "%s()" is deprecated since sonata-project/media-bundle 3.x'
+                .' and will throw a "\TypeError" error in 4.0.',
+                TranslatorInterface::class,
+                __METHOD__
+            ), \E_USER_DEPRECATED);
+        } elseif (!$translator instanceof TranslatorInterface) {
+            throw new \TypeError(sprintf(
+                'Argument 1 passed to "%s()" MUST be an instance of "%s" or "%s", "%s" given.',
+                __METHOD__,
+                LegacyTranslatorInterface::class,
+                TranslatorInterface::class,
+                \get_class($translator)
+            ));
+        }
+
         if (!$security instanceof AuthorizationCheckerInterface) {
```

Comment:
Couldn't we simply add this type declaration in the argument? Or do we need to keep `\InvalidArgumentException` instead of `\TypeError` for some reason?

Reply:
I did not work on this code. IMO this change will be provide BC-break in some cases. Someone can try-catch(InvalidArtgu....).

Reply:
AFAIK, we consider BC the changes in the signature of `__construct()` methods.

Reply:
Scope of this PR is `add translation conracts` and these changes are not in it. Fell free to open new PR for them.